File: my_flutter_app/lib/services/invoice_ledger.dart
/// Invoices and receipts a business writes by hand. Mirrors
/// `functions/invoice_ledger.js`: the same checks the server runs, so the
/// sheet refuses what the callable would refuse, and the same totals, so the
/// phone never shows a number the server would not.
library;

import 'package:intl/intl.dart';

import 'lot_ledger.dart' show lotDateOf;

const invoiceStatusOpen = 'open';
const invoiceStatusPaid = 'paid';

/// How a hand-recorded payment arrived. Same vocabulary as the server.
const invoicePaymentMethods = <String>[
  'cash',
  'zelle',
  'cashapp',
  'venmo',
  'check',
  'card_in_person',
  'other',
];

const invoiceMaxText = 200;
const invoiceMaxNotes = 1000;
const invoiceMaxVin = 17;
const invoiceMaxQuantity = 10000;
const invoiceMaxCents = 100000000;

String _text(Object? v, [int max = invoiceMaxText]) {
  final t = (v ?? '').toString().trim();
  return t.length > max ? t.substring(0, max) : t;
}

int _positiveInt(Object? v) {
  final n = v is num ? v.truncate() : int.tryParse((v ?? '').toString().trim());
  return n != null && n > 0 ? n : 0;
}

int? _cents(Object? v) {
  if (v is num) return v.round();
  return int.tryParse((v ?? '').toString().trim());
}

/// "YYYY-MM-DD" or "" - invoice dates are calendar days with no clock.
String invoiceDayKey(Object? value) {
  if (value == null) return '';
  if (value is String) {
    final m = RegExp(r'^(\d{4})-(\d{2})-(\d{2})').firstMatch(value.trim());
    if (m == null) return '';
    final probe = DateTime.tryParse('${m[1]}-${m[2]}-${m[3]}T00:00:00Z');
    if (probe == null) return '';
    final day = probe.toIso8601String().substring(0, 10);
    return day == '${m[1]}-${m[2]}-${m[3]}' ? day : '';
  }
  final date = lotDateOf(value);
  return date == null ? '' : date.toUtc().toIso8601String().substring(0, 10);
}

String invoiceTodayKey([DateTime? now]) =>
    DateFormat('yyyy-MM-dd').format(now ?? DateTime.now());

// ---------------------------------------------------------------------------
// The invoice.
// ---------------------------------------------------------------------------

class InvoiceDraft {
  const InvoiceDraft({
    this.title = '',
    this.customerName = '',
    this.customerPhone = '',
    this.customerEmail = '',
    this.issuedOn = '',
    this.dueOn = '',
    this.notes = '',
  });

  final String title;
  final String customerName;
  final String customerPhone;
  final String customerEmail;
  final String issuedOn;
  final String dueOn;
  final String notes;
}

/// Error codes. Mirrors `validateInvoice`.
List<String> validateInvoice(InvoiceDraft d) {
  final errors = <String>[];
  if (_text(d.title).isEmpty) errors.add('title_required');
  if (_text(d.customerName).isEmpty) errors.add('customer_name_required');
  final issued = _text(d.issuedOn, 40);
  final issuedDay = invoiceDayKey(issued);
  if (issued.isNotEmpty && issuedDay.isEmpty) errors.add('issued_on_invalid');
  final due = _text(d.dueOn, 40);
  final dueDay = invoiceDayKey(due);
  if (due.isNotEmpty && dueDay.isEmpty) errors.add('due_on_invalid');
  if (issuedDay.isNotEmpty &&
      dueDay.isNotEmpty &&
      dueDay.compareTo(issuedDay) < 0) {
    errors.add('due_before_issued');
  }
  return errors;
}

/// The body `createInvoice` / `updateInvoice` take.
Map<String, dynamic> invoicePayload(InvoiceDraft d) => {
      'title': _text(d.title),
      'customerName': _text(d.customerName),
      'customerPhone': _text(d.customerPhone, 40),
      'customerEmail': _text(d.customerEmail, 180),
      'issuedOn': invoiceDayKey(d.issuedOn),
      'dueOn': invoiceDayKey(d.dueOn),
      'notes': _text(d.notes, invoiceMaxNotes),
    };

// ---------------------------------------------------------------------------
// Lines.
// ---------------------------------------------------------------------------

class InvoiceLineDraft {
  const InvoiceLineDraft({
    this.description = '',
    this.quantity = 1,
    this.unitPriceCents,
    this.vinNumber = '',
  });

  final String description;
  final int quantity;

  /// Null when nothing usable was typed.
  final int? unitPriceCents;
  final String vinNumber;
}

/// "1,250.50" or "$120" as typed → cents, or null when it is not a number.
int? invoiceDollarsToCents(String value) {
  final cleaned = value.replaceAll(RegExp(r'[$,\s]'), '');
  if (cleaned.isEmpty) return null;
  final n = double.tryParse(cleaned);
  return n == null ? null : (n * 100).round();
}

String invoiceCentsToInput(int cents) =>
    cents % 100 == 0 ? '${cents ~/ 100}' : (cents / 100).toStringAsFixed(2);

String _cleanVin(String v) =>
    v.toUpperCase().replaceAll(RegExp(r'[^A-Z0-9]'), '');

/// Error codes. Mirrors `validateInvoiceLine`.
List<String> validateInvoiceLine(InvoiceLineDraft d) {
  final errors = <String>[];
  if (_text(d.description).isEmpty) errors.add('description_required');
  final quantity = _positiveInt(d.quantity);
  if (quantity <= 0) errors.add('quantity_required');
  final unit = d.unitPriceCents;
  if (unit == null || unit < 0) errors.add('unit_price_invalid');
  if (unit != null &&
      unit >= 0 &&
      (quantity > invoiceMaxQuantity ? invoiceMaxQuantity : quantity) * unit >
          invoiceMaxCents) {
    errors.add('amount_too_large');
  }
  final vin = _cleanVin(_text(d.vinNumber, 40));
  if (vin.isNotEmpty && vin.length != invoiceMaxVin) errors.add('vin_invalid');
  return errors;
}

Map<String, dynamic> invoiceLinePayload(InvoiceLineDraft d) {
  final quantity = _positiveInt(d.quantity);
  final vin = _cleanVin(_text(d.vinNumber, 40));
  return {
    'description': _text(d.description),
    'quantity': quantity > invoiceMaxQuantity ? invoiceMaxQuantity : quantity,
    'unitPriceCents': (d.unitPriceCents ?? 0) < 0 ? 0 : (d.unitPriceCents ?? 0),
    'vinNumber': vin.length > invoiceMaxVin ? vin.substring(0, invoiceMaxVin) : vin,
  };
}

// ---------------------------------------------------------------------------
// Payments.
// ---------------------------------------------------------------------------

class InvoicePaymentDraft {
  const InvoicePaymentDraft({
    this.amountCents,
    this.method = 'cash',
    this.paidOn = '',
    this.note = '',
  });

  final int? amountCents;
  final String method;
  final String paidOn;
  final String note;
}

/// Error codes. Mirrors `validateInvoicePayment`.
List<String> validateInvoicePayment(InvoicePaymentDraft d, int balanceCents) {
  final errors = <String>[];
  final amount = d.amountCents;
  if (amount == null || amount <= 0) {
    errors.add('amount_required');
  } else if (amount > invoiceMaxCents) {
    errors.add('amount_too_large');
  } else if (amount > (balanceCents < 0 ? 0 : balanceCents)) {
    errors.add('payment_exceeds_balance');
  }
  if (!invoicePaymentMethods.contains(_text(d.method, 40))) {
    errors.add('payment_method_invalid');
  }
  final paidOn = _text(d.paidOn, 40);
  if (paidOn.isNotEmpty && invoiceDayKey(paidOn).isEmpty) {
    errors.add('paid_on_invalid');
  }
  return errors;
}

Map<String, dynamic> invoicePaymentPayload(InvoicePaymentDraft d) => {
      'amountCents': d.amountCents ?? 0,
      'method': _text(d.method, 40),
      'paidOn': invoiceDayKey(d.paidOn),
      'note': _text(d.note),
    };

// ---------------------------------------------------------------------------
// Stored records.
// ---------------------------------------------------------------------------

class Invoice {
  const Invoice({
    required this.id,
    required this.businessId,
    required this.number,
    required this.title,
    required this.customerName,
    required this.customerPhone,
    required this.customerEmail,
    required this.issuedOn,
    required this.dueOn,
    required this.paidOn,
    required this.notes,
    required this.status,
    required this.totalCents,
    required this.paidCents,
    required this.balanceCents,
    required this.lineCount,
    required this.paymentCount,
    required this.createdByStaffId,
    required this.createdAt,
  });

  final String id;
  final String businessId;
  final String number;
  final String title;
  final String customerName;
  final String customerPhone;
  final String customerEmail;
  final String issuedOn;
  final String dueOn;
  final String paidOn;
  final String notes;
  final String status;
  final int totalCents;
  final int paidCents;
  final int balanceCents;
  final int lineCount;
  final int paymentCount;
  final String createdByStaffId;
  final DateTime? createdAt;

  bool get isPaid => status == invoiceStatusPaid;
  bool get isOpen => !isPaid;

  /// "INV-0007 · Corolla".
  String get displayName =>
      [number, title].where((p) => p.isNotEmpty).join(' · ');

  factory Invoice.fromMap(String id, Map<String, dynamic> d) => Invoice(
        id: id,
        businessId: _text(d['businessId']),
        number: _text(d['number'], 20),
        title: _text(d['title']),
        customerName: _text(d['customerName']),
        customerPhone: _text(d['customerPhone'], 40),
        customerEmail: _text(d['customerEmail'], 180),
        issuedOn: invoiceDayKey(d['issuedOn']),
        dueOn: invoiceDayKey(d['dueOn']),
        paidOn: invoiceDayKey(d['paidOn']),
        notes: _text(d['notes'], invoiceMaxNotes),
        status: _text(d['status'], 20) == invoiceStatusPaid
            ? invoiceStatusPaid
            : invoiceStatusOpen,
        totalCents: _cents(d['totalCents']) ?? 0,
        paidCents: _cents(d['paidCents']) ?? 0,
        balanceCents: _cents(d['balanceCents']) ?? 0,
        lineCount: _positiveInt(d['lineCount']),
        paymentCount: _positiveInt(d['paymentCount']),
        createdByStaffId: _text(d['createdByStaffId']),
        createdAt: lotDateOf(d['createdAt']),
      );
}

class InvoiceLine {
  const InvoiceLine({
    required this.id,
    required this.invoiceId,
    required this.description,
    required this.quantity,
    required this.unitPriceCents,
    required this.amountCents,
    required this.vinNumber,
    required this.addedByStaffId,
    required this.createdAt,
  });

  final String id;
  final String invoiceId;
  final String description;
  final int quantity;
  final int unitPriceCents;
  final int amountCents;
  final String vinNumber;
  final String addedByStaffId;
  final DateTime? createdAt;

  factory InvoiceLine.fromMap(String id, Map<String, dynamic> d) => InvoiceLine(
        id: id,
        invoiceId: _text(d['invoiceId']),
        description: _text(d['description']),
        quantity: _positiveInt(d['quantity']) == 0 ? 1 : _positiveInt(d['quantity']),
        unitPriceCents: _cents(d['unitPriceCents']) ?? 0,
        amountCents: _cents(d['amountCents']) ?? 0,
        vinNumber: _cleanVin(_text(d['vinNumber'], 40)),
        addedByStaffId: _text(d['addedByStaffId']),
        createdAt: lotDateOf(d['createdAt']),
      );
}

class InvoicePayment {
  const InvoicePayment({
    required this.id,
    required this.invoiceId,
    required this.amountCents,
    required this.method,
    required this.paidOn,
    required this.note,
    required this.reverted,
    required this.receivedByStaffId,
    required this.createdAt,
  });

  final String id;
  final String invoiceId;
  final int amountCents;
  final String method;
  final String paidOn;
  final String note;
  final bool reverted;
  final String receivedByStaffId;
  final DateTime? createdAt;

  factory InvoicePayment.fromMap(String id, Map<String, dynamic> d) =>
      InvoicePayment(
        id: id,
        invoiceId: _text(d['invoiceId']),
        amountCents: _cents(d['amountCents']) ?? 0,
        method: _text(d['method'], 40),
        paidOn: invoiceDayKey(d['paidOn']),
        note: _text(d['note']),
        reverted: d['reverted'] == true,
        receivedByStaffId: _text(d['receivedByStaffId']),
        createdAt: lotDateOf(d['createdAt']),
      );
}

// ---------------------------------------------------------------------------
// What it adds up to.
// ---------------------------------------------------------------------------

class InvoiceTotals {
  const InvoiceTotals({
    required this.totalCents,
    required this.paidCents,
    required this.balanceCents,
    required this.lineCount,
    required this.paymentCount,
    required this.status,
  });

  final int totalCents;
  final int paidCents;
  final int balanceCents;
  final int lineCount;
  final int paymentCount;
  final String status;
}

/// Mirrors `invoiceTotals`: total minus the payments that still stand; paid
/// when nothing is owed - but a blank invoice is never paid.
InvoiceTotals invoiceTotals(
    Iterable<InvoiceLine> lines, Iterable<InvoicePayment> payments) {
  final rows = lines.toList();
  final standing = payments.where((p) => !p.reverted).toList();
  final total = rows.fold<int>(0, (s, l) => s + (l.amountCents < 0 ? 0 : l.amountCents));
  final paid = standing.fold<int>(0, (s, p) => s + (p.amountCents < 0 ? 0 : p.amountCents));
  final balance = total - paid < 0 ? 0 : total - paid;
  return InvoiceTotals(
    totalCents: total,
    paidCents: paid,
    balanceCents: balance,
    lineCount: rows.length,
    paymentCount: standing.length,
    status: rows.isNotEmpty && balance == 0 ? invoiceStatusPaid : invoiceStatusOpen,
  );
}

/// Overdue is a fact about an open invoice with a due day behind today.
bool invoiceIsOverdue(Invoice invoice, [String? today]) {
  final t = today ?? invoiceTodayKey();
  return invoice.isOpen && invoice.dueOn.isNotEmpty && invoice.dueOn.compareTo(t) < 0;
}

/// "$1,250.50", never rounded away.
String invoiceMoney(int cents) {
  final whole = NumberFormat('#,##0', 'en_US').format(cents.abs() ~/ 100);
  final part = (cents.abs() % 100).toString().padLeft(2, '0');
  return '${cents < 0 ? '-' : ''}\$$whole.$part';
}

/// Newest invoice date first, then by number, so today's work is on top.
List<Invoice> sortInvoices(Iterable<Invoice> rows) => rows.toList()
  ..sort((a, b) {
    final byDay = b.issuedOn.compareTo(a.issuedOn);
    return byDay != 0 ? byDay : b.number.compareTo(a.number);
  });

const invoiceFilterAll = 'all';
const invoiceFilterOpen = 'open';
const invoiceFilterOverdue = 'overdue';
const invoiceFilterPaid = 'paid';
const invoiceFilters = <String>[
  invoiceFilterOpen,
  invoiceFilterOverdue,
  invoiceFilterPaid,
  invoiceFilterAll,
];

List<Invoice> filterInvoices(
  Iterable<Invoice> rows,
  String filter,
  String search, [
  String? today,
]) {
  final q = search.trim().toLowerCase();
  return [
    for (final row in rows)
      if ((filter != invoiceFilterOpen || row.isOpen) &&
          (filter != invoiceFilterPaid || row.isPaid) &&
          (filter != invoiceFilterOverdue || invoiceIsOverdue(row, today)) &&
          (q.isEmpty ||
              '${row.number} ${row.title} ${row.customerName} ${row.customerPhone}'
                  .toLowerCase()
                  .contains(q)))
        row,
  ];
}

/// The message pasted into WhatsApp beside the PDF. Mirrors the server's.
String invoiceTextSummary({
  required Invoice invoice,
  required List<InvoiceLine> lines,
  required List<InvoicePayment> payments,
  required String businessName,
}) {
  final totals = invoiceTotals(lines, payments);
  final kind = totals.status == invoiceStatusPaid ? 'Receipt' : 'Invoice';
  final out = <String>[
    '${businessName.trim().isEmpty ? 'Invoice' : businessName.trim()} — $kind'
        '${invoice.number.isNotEmpty ? ' ${invoice.number}' : ''}',
    '${invoice.title}${invoice.issuedOn.isNotEmpty ? ' · ${invoice.issuedOn}' : ''}',
    'For: ${invoice.customerName.isEmpty ? '—' : invoice.customerName}',
    '',
  ];
  for (final l in lines) {
    final each = l.quantity > 1
        ? ' (${l.quantity} × ${invoiceMoney(l.unitPriceCents)})'
        : '';
    out.add('${l.description}$each — ${invoiceMoney(l.amountCents)}'
        '${l.vinNumber.isNotEmpty ? '\nVIN ${l.vinNumber}' : ''}');
  }
  out
    ..add('')
    ..add('Total: ${invoiceMoney(totals.totalCents)}');
  if (totals.paidCents > 0) out.add('Paid: -${invoiceMoney(totals.paidCents)}');
  out.add(totals.balanceCents > 0
      ? 'BALANCE DUE: ${invoiceMoney(totals.balanceCents)}'
          '${invoice.dueOn.isNotEmpty ? ' (due ${invoice.dueOn})' : ''}'
      : 'PAID IN FULL');
  return out.join('\n');
}

File: admin_web/src/lib/invoice-ledger.ts
/**
 * Invoices and receipts a business writes by hand. Mirrors
 * `my_flutter_app/functions/invoice_ledger.js`: the same checks the server
 * runs, so the form refuses what the callable would refuse, and the same
 * totals, so the page never shows a number the server would not.
 */

export type InvoiceStatus = "open" | "paid";

export const INVOICE_PAYMENT_METHODS = [
  "cash", "zelle", "cashapp", "venmo", "check", "card_in_person", "other",
] as const;
export type InvoicePaymentMethod = (typeof INVOICE_PAYMENT_METHODS)[number];

export const INVOICE_PAYMENT_METHOD_LABELS: Record<InvoicePaymentMethod, string> = {
  cash: "Cash",
  zelle: "Zelle",
  cashapp: "Cash App",
  venmo: "Venmo",
  check: "Check",
  card_in_person: "Card in person",
  other: "Other",
};

const MAX_TEXT = 200;
const MAX_NOTES = 1000;
const MAX_VIN = 17;
const MAX_QUANTITY = 10000;
export const MAX_INVOICE_CENTS = 100000000;

export const INVOICE_MESSAGES = {
  title_required: "Give the invoice a short title (what it is for).",
  customer_name_required: "Say who the invoice is for.",
  issued_on_invalid: "The invoice date is not a real date.",
  due_on_invalid: "The due date is not a real date.",
  due_before_issued: "The due date is before the invoice date.",
  description_required: "Say what the line is.",
  quantity_required: "Enter how many (at least one).",
  unit_price_invalid: "Enter the price for one.",
  amount_too_large: "That amount is larger than an invoice can carry.",
  vin_invalid: "A VIN is 17 letters and digits.",
  amount_required: "Enter the amount received.",
  payment_method_invalid: "Say how the payment arrived.",
  paid_on_invalid: "The payment date is not a real date.",
  payment_exceeds_balance: "That is more than what is still owed.",
  invoice_not_found: "That invoice no longer exists.",
  line_not_found: "That line no longer exists.",
  payment_not_found: "That payment no longer exists.",
  payment_already_reverted: "That payment was already reverted.",
  invoice_has_lines: "Remove the lines and payments before deleting.",
} as const;
export type InvoiceError = keyof typeof INVOICE_MESSAGES;

type Row = Record<string, unknown>;
const asRow = (value: unknown): Row =>
  value && typeof value === "object" ? (value as Row) : {};
const text = (value: unknown, max = MAX_TEXT) =>
  String(value ?? "").trim().slice(0, max);
const positiveInt = (value: unknown) => {
  const n = Math.trunc(Number(value));
  return Number.isFinite(n) && n > 0 ? n : 0;
};
const cents = (value: unknown) => {
  const n = Math.round(Number(value));
  return Number.isFinite(n) ? n : NaN;
};

/** "YYYY-MM-DD" or "" — invoice dates are calendar days with no clock. */
export function invoiceDayKey(value: unknown): string {
  if (!value) return "";
  if (typeof value === "string") {
    const match = /^(\d{4})-(\d{2})-(\d{2})/.exec(value.trim());
    if (!match) return "";
    const probe = new Date(`${match[1]}-${match[2]}-${match[3]}T00:00:00Z`);
    if (Number.isNaN(probe.getTime())) return "";
    const day = probe.toISOString().slice(0, 10);
    return day === `${match[1]}-${match[2]}-${match[3]}` ? day : "";
  }
  const withToDate = value as { toDate?: () => Date };
  const date = typeof withToDate.toDate === "function"
    ? withToDate.toDate()
    : new Date(value as string | number | Date);
  return Number.isNaN(date.getTime()) ? "" : date.toISOString().slice(0, 10);
}

export function todayKey(): string {
  const now = new Date();
  const m = String(now.getMonth() + 1).padStart(2, "0");
  const d = String(now.getDate()).padStart(2, "0");
  return `${now.getFullYear()}-${m}-${d}`;
}

export function invoiceMessage(codes: readonly string[]): string {
  return codes.map((c) => (c in INVOICE_MESSAGES ? INVOICE_MESSAGES[c as InvoiceError] : c)).join(" ");
}

// ---------------------------------------------------------------------------
// The invoice.
// ---------------------------------------------------------------------------

export type InvoiceDraft = {
  title: string;
  customerName: string;
  customerPhone: string;
  customerEmail: string;
  issuedOn: string;
  dueOn: string;
  notes: string;
};

export const emptyInvoiceDraft: InvoiceDraft = {
  title: "",
  customerName: "",
  customerPhone: "",
  customerEmail: "",
  issuedOn: "",
  dueOn: "",
  notes: "",
};

export function invoiceDraftFromRow(row: Row): InvoiceDraft {
  return {
    title: text(row.title),
    customerName: text(row.customerName),
    customerPhone: text(row.customerPhone, 40),
    customerEmail: text(row.customerEmail, 180),
    issuedOn: invoiceDayKey(row.issuedOn),
    dueOn: invoiceDayKey(row.dueOn),
    notes: text(row.notes, MAX_NOTES),
  };
}

/** Same checks as the server's `validateInvoice`. */
export function validateInvoiceDraft(draft: InvoiceDraft): InvoiceError[] {
  const errors: InvoiceError[] = [];
  if (!text(draft.title)) errors.push("title_required");
  if (!text(draft.customerName)) errors.push("customer_name_required");
  const issued = text(draft.issuedOn, 40);
  const issuedDay = invoiceDayKey(issued);
  if (issued && !issuedDay) errors.push("issued_on_invalid");
  const due = text(draft.dueOn, 40);
  const dueDay = invoiceDayKey(due);
  if (due && !dueDay) errors.push("due_on_invalid");
  if (issuedDay && dueDay && dueDay < issuedDay) errors.push("due_before_issued");
  return errors;
}

export function invoicePayload(draft: InvoiceDraft) {
  return {
    title: text(draft.title),
    customerName: text(draft.customerName),
    customerPhone: text(draft.customerPhone, 40),
    customerEmail: text(draft.customerEmail, 180),
    issuedOn: invoiceDayKey(draft.issuedOn),
    dueOn: invoiceDayKey(draft.dueOn),
    notes: text(draft.notes, MAX_NOTES),
  };
}

// ---------------------------------------------------------------------------
// Lines.
// ---------------------------------------------------------------------------

export type InvoiceLineDraft = {
  description: string;
  /** As typed; defaults to one. */
  quantity: string;
  /** In dollars as typed; sent in cents. */
  unitPrice: string;
  vinNumber: string;
};

export const emptyInvoiceLineDraft: InvoiceLineDraft = {
  description: "",
  quantity: "1",
  unitPrice: "",
  vinNumber: "",
};

export function invoiceLineDraftFromRow(row: Row): InvoiceLineDraft {
  return {
    description: text(row.description),
    quantity: String(positiveInt(row.quantity) || 1),
    unitPrice: centsToInput(row.unitPriceCents),
    vinNumber: text(row.vinNumber, 40),
  };
}

export function dollarsToCents(value: string): number {
  const cleaned = String(value ?? "").replace(/[$,\s]/g, "");
  if (!cleaned) return NaN;
  const n = Number(cleaned);
  return Number.isFinite(n) ? Math.round(n * 100) : NaN;
}

export function centsToInput(value: unknown): string {
  const n = cents(value);
  if (!Number.isFinite(n)) return "";
  return n % 100 === 0 ? String(n / 100) : (n / 100).toFixed(2);
}

/** Same checks as the server's `validateInvoiceLine`. */
export function validateInvoiceLineDraft(draft: InvoiceLineDraft): InvoiceError[] {
  const errors: InvoiceError[] = [];
  if (!text(draft.description)) errors.push("description_required");
  const quantity = positiveInt(draft.quantity);
  if (quantity <= 0) errors.push("quantity_required");
  const unit = dollarsToCents(draft.unitPrice);
  if (!Number.isFinite(unit) || unit < 0) errors.push("unit_price_invalid");
  if (Number.isFinite(unit) && unit >= 0 && Math.min(quantity, MAX_QUANTITY) * unit > MAX_INVOICE_CENTS) {
    errors.push("amount_too_large");
  }
  const vin = text(draft.vinNumber, 40).toUpperCase().replace(/[^A-Z0-9]/g, "");
  if (vin && vin.length !== MAX_VIN) errors.push("vin_invalid");
  return errors;
}

export function invoiceLinePayload(draft: InvoiceLineDraft) {
  return {
    description: text(draft.description),
    quantity: Math.min(MAX_QUANTITY, positiveInt(draft.quantity) || 1),
    unitPriceCents: Math.max(0, dollarsToCents(draft.unitPrice) || 0),
    vinNumber: text(draft.vinNumber, 40).toUpperCase().replace(/[^A-Z0-9]/g, "").slice(0, MAX_VIN),
  };
}

// ---------------------------------------------------------------------------
// Payments.
// ---------------------------------------------------------------------------

export type InvoicePaymentDraft = {
  amount: string;
  method: InvoicePaymentMethod;
  paidOn: string;
  note: string;
  /** One of the invoice's line ids, or "" for the invoice as a whole. */
  forLineId: string;
};

export const emptyInvoicePaymentDraft: InvoicePaymentDraft = {
  amount: "",
  method: "cash",
  paidOn: "",
  note: "",
  forLineId: "",
};

/** Same checks as the server's `validateInvoicePayment`. */
export function validateInvoicePaymentDraft(draft: InvoicePaymentDraft, balanceCents: number): InvoiceError[] {
  const errors: InvoiceError[] = [];
  const amount = dollarsToCents(draft.amount);
  if (!Number.isFinite(amount) || amount <= 0) errors.push("amount_required");
  else if (amount > MAX_INVOICE_CENTS) errors.push("amount_too_large");
  else if (amount > Math.max(0, Number(balanceCents) || 0)) errors.push("payment_exceeds_balance");
  if (!(INVOICE_PAYMENT_METHODS as readonly string[]).includes(text(draft.method, 40))) {
    errors.push("payment_method_invalid");
  }
  const paidOn = text(draft.paidOn, 40);
  if (paidOn && !invoiceDayKey(paidOn)) errors.push("paid_on_invalid");
  return errors;
}

export function invoicePaymentPayload(draft: InvoicePaymentDraft) {
  return {
    amountCents: dollarsToCents(draft.amount),
    method: text(draft.method, 40),
    paidOn: invoiceDayKey(draft.paidOn),
    note: text(draft.note),
    forLineId: text(draft.forLineId, 200),
  };
}

/** How a payment reads in a list or on the paper. Mirrors the server's. */
export function invoicePaymentLabel(payment: Row, methodLabel: (m: string) => string = (m) => m): string {
  return [
    invoiceDayKey(payment.paidOn),
    text(payment.method, 40) ? methodLabel(text(payment.method, 40)) : "",
    text(payment.forDescription) ? `for ${text(payment.forDescription)}` : "",
    text(payment.note),
  ].filter(Boolean).join(" · ");
}

// ---------------------------------------------------------------------------
// What it adds up to.
// ---------------------------------------------------------------------------

export type InvoiceTotals = {
  totalCents: number;
  paidCents: number;
  balanceCents: number;
  lineCount: number;
  paymentCount: number;
  status: InvoiceStatus;
};

export function invoiceTotals(lines: readonly Row[], payments: readonly Row[]): InvoiceTotals {
  const paid = payments.filter((p) => p.reverted !== true);
  const totalCents = lines.reduce((sum, l) => sum + Math.max(0, cents(l.amountCents) || 0), 0);
  const paidCents = paid.reduce((sum, p) => sum + Math.max(0, cents(p.amountCents) || 0), 0);
  const balanceCents = Math.max(0, totalCents - paidCents);
  return {
    totalCents,
    paidCents,
    balanceCents,
    lineCount: lines.length,
    paymentCount: paid.length,
    status: lines.length > 0 && balanceCents === 0 ? "paid" : "open",
  };
}

export function invoiceStatus(row: Row): InvoiceStatus {
  return text(row.status, 20) === "paid" ? "paid" : "open";
}

export function invoiceIsOverdue(row: Row, today = todayKey()): boolean {
  const due = invoiceDayKey(row.dueOn);
  return invoiceStatus(row) === "open" && Boolean(due) && due < today;
}

/** "Receipt" once settled, else "Invoice". */
export function invoiceKind(row: Row): "Invoice" | "Receipt" {
  return invoiceStatus(row) === "paid" ? "Receipt" : "Invoice";
}

export function invoiceTitle(row: Row): string {
  const number = text(row.number, 20);
  const title = text(row.title);
  return [number, title].filter(Boolean).join(" · ") || "Invoice";
}

export function moneyText(value: unknown): string {
  const n = Math.round(Number(value) || 0);
  const whole = Math.trunc(Math.abs(n) / 100);
  const part = String(Math.abs(n) % 100).padStart(2, "0");
  return `${n < 0 ? "-" : ""}$${whole.toLocaleString("en-US")}.${part}`;
}

export type InvoiceFilter = "" | "open" | "paid" | "overdue";

export function filterInvoices(rows: readonly Row[], filter: InvoiceFilter, search: string, today = todayKey()): Row[] {
  const q = text(search, 120).toLowerCase();
  return rows.filter((row) => {
    if (filter === "open" && invoiceStatus(row) !== "open") return false;
    if (filter === "paid" && invoiceStatus(row) !== "paid") return false;
    if (filter === "overdue" && !invoiceIsOverdue(row, today)) return false;
    if (!q) return true;
    const hay = [row.number, row.title, row.customerName, row.customerPhone]
      .map((v) => text(v).toLowerCase()).join(" ");
    return hay.includes(q);
  });
}

/** Newest invoice date first, then by number, so today's work is on top. */
export function sortInvoices(rows: readonly Row[]): Row[] {
  return [...rows].sort((a, b) => {
    const byDay = invoiceDayKey(b.issuedOn).localeCompare(invoiceDayKey(a.issuedOn));
    if (byDay !== 0) return byDay;
    return text(b.number, 20).localeCompare(text(a.number, 20));
  });
}

/** The message pasted into WhatsApp beside the PDF. Mirrors the server's. */
export function invoiceTextSummary(input: {
  invoice: Row;
  lines: readonly Row[];
  payments: readonly Row[];
  businessName: string;
}): string {
  const inv = input.invoice;
  const totals = invoiceTotals(input.lines, input.payments);
  const number = text(inv.number, 20);
  const out = [
    `${text(input.businessName) || "Invoice"} — ${totals.status === "paid" ? "Receipt" : "Invoice"}${number ? ` ${number}` : ""}`,
    `${text(inv.title)}${invoiceDayKey(inv.issuedOn) ? ` · ${invoiceDayKey(inv.issuedOn)}` : ""}`,
    `For: ${text(inv.customerName) || "—"}`,
    "",
  ];
  for (const l of input.lines) {
    const q = positiveInt(l.quantity) || 1;
    const each = q > 1 ? ` (${q} × ${moneyText(l.unitPriceCents)})` : "";
    const vin = text(l.vinNumber, 17);
    out.push(`${text(l.description)}${each} — ${moneyText(l.amountCents)}${vin ? `\nVIN ${vin}` : ""}`);
  }
  out.push("", `Total: ${moneyText(totals.totalCents)}`);
  if (totals.paidCents > 0) {
    out.push(`Paid: -${moneyText(totals.paidCents)}`);
    for (const p of input.payments.filter((x) => x.reverted !== true)) {
      out.push(`  ${invoicePaymentLabel(p)} — ${moneyText(p.amountCents)}`);
    }
  }
  const due = invoiceDayKey(inv.dueOn);
  out.push(totals.balanceCents > 0
    ? `BALANCE DUE: ${moneyText(totals.balanceCents)}${due ? ` (due ${due})` : ""}`
    : "PAID IN FULL");
  return out.join("\n");
}

export function invoiceCallableFailure(error: unknown): string {
  const e = asRow(error);
  return text(e.message, 500) || "The change did not save.";
}
